src/main/java/io/confluent/connect/jdbc/source/SbdBulkTableQuerier.java

```diff
@@ -0,0 +1,250 @@
+        currentBatchId = batchId;
+        currentBatchTime = batchTime;
+      }
+    }
+
+    if (currentBatchId != null) {
+      log.info("Configured to retry batch.id={}, batch.time={}", currentBatchId, currentBatchTime);
+    }
+  }
+
+  public long getLastUpdate() {
+    return lastUpdate;
+  }
+
+  protected ResultSet executeQuery() throws SQLException {
+    return stmt.executeQuery();
+  }
+
+  @Override
+  protected void createPreparedStatement(Connection db) throws SQLException {
+    ExpressionBuilder builder = dialect.expressionBuilder();
+    switch (mode) {
+      case TABLE:
+        builder.append("SELECT * FROM ").append(tableId);
+
+        break;
+      case QUERY:
+        builder.append(query);
+
+        break;
+      default:
+        throw new ConnectException("Unknown mode: " + mode);
+    }
+
+    addSuffixIfPresent(builder);
+
+    String queryStr = builder.toString();
+
+    recordQuery(queryStr);
+    log.debug("{} prepared SQL query: {}", this, queryStr);
+    stmt = dialect.createPreparedStatement(db, queryStr, ResultSet.TYPE_SCROLL_INSENSITIVE);
+  }
+
+  @Override
+  public SourceRecord extractRecord() throws SQLException {
+    SourceRecord sr = super.extractRecord();
+
+    final int rowIndex = resultSet.getRow(); // 1 for first row, 2 for second one
+    final boolean isLastRecord = resultSet.isLast();
+
+    // we retry failing batches during 3h but not if we are after the next scheduled execution time
+    if (resultSet.isFirst()) {
+      Instant now = Instant.ofEpochMilli(time.milliseconds());
+      if (
+          now.minus(3, ChronoUnit.HOURS).isAfter(Instant.ofEpochMilli(lastUpdate))
+          && now.isBefore(Instant.ofEpochMilli(nextExecutionTime))
+      ) {
+        currentBatchId = null;
+      }
+    }
+
+    // new batch
+    if (currentBatchId == null) {
+      currentBatchId = UUID.randomUUID().toString();
+      currentBatchTime = Instant.ofEpochMilli(currentExecutionTime);
+      currentBatchStartTime = currentBatchTime;
+
+      // determine the size of the batch
+      assert resultSet.isFirst();
+      resultSet.last();
+      batchSize = resultSet.getRow();
+      resultSet.first();
+    } else if (currentBatchStartTime == null) {
+      // batch already defined, here it's a new attempt
+      currentBatchStartTime = Instant.ofEpochMilli(time.milliseconds());
+    }
+
+    Headers headers = new ConnectHeaders(sr.headers());
+    headers.add(
+        HEADER_BATCH_ID,
+        new SchemaAndValue(Schema.STRING_SCHEMA, currentBatchId)
+    );
+    headers.add(
+        HEADER_BATCH_TIME,
+        new SchemaAndValue(Schema.STRING_SCHEMA, currentBatchTime.toString())
+    );
+    headers.add(
+        HEADER_BATCH_SIZE,
+        new SchemaAndValue(Schema.INT32_SCHEMA, batchSize)
+    );
+    headers.add(
+        HEADER_BATCH_INDEX,
+        new SchemaAndValue(Schema.INT32_SCHEMA, rowIndex)
+    );
+    headers.add(
+        HEADER_BATCH_STARTED_AT,
```

Comment:
do you need it on every single record? Maybe adding it to the first one is enough.